src/cockpit/beiboot.py

```diff
@@ -165,12 +166,21 @@ async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[st
             # Let's avoid all of that by just showing nothing.
             return None
 
+        # is this a host key prompt?
+        fp_match = re.search(r'\n(\w+) key fingerprint is ([^.]+)\.', prompt)
+        args = {}
+        if fp_match:
+            args['host-key'] = f'{fp_match.group(2)} {fp_match.group(1)}'
```

Comment:
I guess this isn't quite right — this is the fingerprint, not the key...

Reply:
Hm yes, but we don't ever get the actual key via SSH_ASKPASS; I figure that's why you went through this whole business of `KnownHostCommand`? Do we need to use that after all?

Reply:
I.e. I just consider that as the "new key format" -- key fingerprints are just the "language" that `ssh` speaks, while for `cockpit-ssh` it was the full keys. But does that matter? Or what would you suggest?

Reply:
I guess this is still unresolved — I worry about existing users who might be concerned that their key no longer matches, on upgrade.  Do we have a test for that?

Reply:
(don't get me wrong — I think this change is good and ultimately necessary if we're to abandon our dependency on hostkey commands... but I feel like we need to try a bit harder on the migration front...)

Reply:
The ultimate/persistent place to store that information is ~/.ssh/known_hosts, both for cockpit-ssh and cockpit-beiboot. That format is fixed, and ssh/cockpit-ssh always store the full keys there. As far as I can see this only affects the wire protocol between cockpit-{ssh,beiboot} and the UI.

What migration would/could we possibly do here? To be honest I don't even have an idea how to "test" this against the old format -- we either use cockpit-ssh or cockpit-beiboot, and the UI just needs to get along with either. We never showed full keys in the UI, always just fingerprints.

Reply:
It's been a while, so need to page that stuff back into my brain. Login page currently stores the full host key in localStorage:
```
known_hosts:"{"[127.0.0.2]:2201":"[127.0.0.2]:2201 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBERg5Z5mVTMkmOaIICuFnrLZs/DWPFq/PvNkrady3sxvETrtyF58yjYfIt9iOL6Tcve65F+iDYUsyv9r431ToM=\n"}"
```

The corresponding fingerprint is "SHA256:sq4xrPrvIQ7MChHi8fZedhXQxqqba3BDLsuSOu0/SPc"

It's awkward to call `ssh-keygen` on this on the login page, as at that point we don't even have a bridge. But that fingerprint calculation isn't obvious -- the most straightforward computation gives something else:

```
# echo -n 'AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBERg5Z5mVTMkmOaIICuFnrLZs/DWPFq/PvNkrady3sxvETrtyF58yjYfIt9iOL6Tcve65F+iDYUsyv9r431ToM=' | base64 -d | sha256sum -
b2ae31acfaef210ecc0a11e2f1f65e7615d0c6aa9b6b70432ecb923aed3f48f7  -
```

I also tried to directly SHA the base64 encoded key, but it's also different.

I found a [stackexchange question](https://security.stackexchange.com/questions/152428/how-is-an-ssh-public-key-fingerprint-calculated), and re-tried with Python and that *almost* works:
```
>>> import hashlib, base64
>>> b64pubkey = 'AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBERg5Z5mVTMkmOaIICuFnrLZs/DWPFq/PvNkrady3sxvETrtyF58yjYfIt9iOL6Tcve65F+iDYUsyv9r431ToM='
>>> s = hashlib.sha256()
>>> s.update(base64.b64decode(b64pubkey))
>>> h = s.digest()
>>> h
b'\xb2\xae1\xac\xfa\xef!\x0e\xcc\n\x11\xe2\xf1\xf6^v\x15\xd0\xc6\xaa\x9bkpC.\xcb\x92:\xed?H\xf7'
>>> base64.b64encode(h)
b'sq4xrPrvIQ7MChHi8fZedhXQxqqba3BDLsuSOu0/SPc='
```

The difference is the extra `=` padding at the end. We could ignore these for the comparison.

Reply:
If the key type doesn't match, show the "unknown host", not the "warning danger host key changed" dialog.

Reply:
I reworked this. The integration test makes relatively strong assumptions, let's see how far I get with that. I may have to weaken it a bit (make more dynamic).

Reply:
This is mostly working now, everything uses full host keys. See the details and caveat in the commit message.